### PR TITLE
adds create container with PUT and more

### PR DESCRIPTION
**Pull request by bourgeoa**
@michielbdejong 
adds create container with PUT and some updates to be solid/specification compliant, and pass solid-crud-tests :
- allow creation of container with POST and PUT
- contentType is a MUST with POST, PUT, PATCH. So fail and do not default if not exists, but default to mime-type default if unknown (`application/octet-stream`)
- add controls to check that a folder and file cannot have same name excluding / with POST, PUT and PATCH
- with PUT, PATCH delete existing file if : with same URI, newly creating file is different from existing file (delete test.txt$.ttl if creating test.txt)
- modification to pubsub parent to pass solid-crud-tests
- delete root and delete root/.acl are forbidden
- no POST for auxiliary resources

### CODE REVIEW

**Comment by bourgeoa** [comment]
Squashed and merge to the `create-container` branch to create a clean commit and have a test branch.

File: lib/ldp.js [hunk 1]
```diff
@@ -210,18 +239,17 @@ class LDP {
...
-      throw error(409,
-        'PUT not supported on containers, use POST instead')
-    }
+    const container = (url.url || url).endsWith('/')
 
-    // PUT without content type is forbidden
-    if (!contentType) {
-      throw error(415,
-        'PUT request require a valid content type via the Content-Type header')
+    // PUT without content type is forbidden. Except for container
```
Comment (TallTed) [inline lib/ldp.js:244]:
```suggestion
    // PUT without content type is forbidden, unless PUTting container
```

File: test/integration/http-test.js [hunk 2]
```diff
@@ -478,9 +478,28 @@ describe('HTTP APIs', function () {
...
+    })
+    it('should create new resource and delete old path if different', function (done) {
+      server.put('/put-resource-1.ttl')
+        .send(putRequestBody)
+        .set('content-type', 'text/turtle')
+        .expect(201)
+        .end(function (err) {
+          if (err) return done(err)
+          if (fs.existsSync(path.join(__dirname, '../resources/put-resource-1.ttl$.txt'))) {
+            return done(new Error('Can read old file that should be deleted'))
```
Comment (TallTed) [inline test/integration/http-test.js:498]:
```suggestion
            return done(new Error('Can read old file that should have been deleted'))
```

File: test/integration/http-test.js [hunk 3]
```diff
@@ -613,24 +702,47 @@ describe('HTTP APIs', function () {
         .expect('location', /.*\.ttl/)
         .expect(201, done)
     })
-    it('should error with 415 if the body is empty and no content type is provided', function (done) {
+    it('should error with 404 to create folder with same name than a resource', function (done) {
```
Comment (TallTed) [inline test/integration/http-test.js:705]:
```suggestion
    it('should error with 404 to create folder with same name as a resource', function (done) {
```

File: test/integration/http-test.js [hunk 4]
```diff
@@ -495,11 +514,51 @@ describe('HTTP APIs', function () {
         .expect(hasHeader('acl', 'baz.ttl' + suffixAcl))
         .expect(201, done)
     })
-    it('should return 409 code when trying to put to a container',
+    it('should not create new resource if folder with same name', function (done) {
```
Comment (TallTed) [inline test/integration/http-test.js:517]:
```suggestion
    it('should not create new resource if folder with same name exists', function (done) {
```

File: test/integration/http-test.js [hunk 5]
```diff
@@ -683,25 +795,41 @@ describe('HTTP APIs', function () {
...
         })
     })
     it('should be able to access newly container', function (done) {
-      server.get('/post-tests/loans/')
+      server.get('/post-tests/loans.ttl/')
         .expect('content-type', /text\/turtle/)
         .expect(200, done)
     })
-
+    it('should error with 404 to create resource with same name than a container', function (done) {
```
Comment (TallTed) [inline test/integration/http-test.js:816]:
```suggestion
    it('should error with 404 to create resource with same name as a container', function (done) {
```

**Comment by bourgeoa** [comment]
Thank all comments taken

**Pull request merged**